File: cppversion/TimeBasedRecommendor/TimeBasedRecommendor/Candidate.cpp
#include "stdafx.h"
#include "Candidate.h"
#include "string.h"
using namespace std;

Candidate::Candidate(Name title)
{
	initialize();
	this->name = title;
}
Candidate::Candidate(void)
{
	//initialize();
}
// currently the copy constructor does not copy ratings because a candidate with ratings should never be copied
/*Candidate::Candidate(const Candidate& original)
{
	this->name = original.name;
	initialize();
}*/
void Candidate::setName(Name newName)
{
	//printf("setting candidate name %s \r\n", newName.getName().c_str());
	this->name = newName;
	initialize();
}
Name Candidate::getName(void)
{
	return this->name;
}
void Candidate::addParent(Name newName)
{
	//printf("adding parent name %s \r\n", newName.getName().c_str());
	this->parentNames.push_back(newName);
	//this->parentName = newName;
}
void Candidate::addParent(Candidate* newCandidate)
{
	//printf("adding parent name %s \r\n", newName.getName().c_str());
	this->parents.push_back(newCandidate);
	//this->parentName = newName;
}
void Candidate::addChild(Candidate* newChild)
{
	this->children.push_back(newChild);
}

std::vector<Name>* Candidate::getParentNames(void)
{
	return &(this->parentNames);
}
std::vector<Candidate*>* Candidate::getParents(void)
{
	return &(this->parents);
}

std::vector<Candidate*>* Candidate::getChildren(void)
{
	return &(this->children);
}


void Candidate::giveRating(Rating rating)
{
	unsigned int i;
	// update the moving averages of the ratings
	for (i = 0; i < this->ratingEstimators.size(); i++)
	{
		ratingEstimators[i].addRating(rating);
	}
	this->actualRatingHistory.addRating(rating);
}
// inform the Candidate that it was listened to during a certain interval
void Candidate::giveParticipation(Participation participation)
{
	unsigned int i;
	// update the moving averages of the frequencies
	/*if (numRatings > 0)
	{
		// keep track of the interval of inactivity between the previous and current rating intervals
		Participation idleTime;
		idleTime.setStartTime(latestRatingTime);
		idleTime.setEndTime(participation.getStartTime());
		idleTime.setIntensity(0);
		
		for (i = 0; i < this->frequencyEstimators.size(); i++)
		{
			frequencyEstimators[i].rateInterval(idleTime);
		}
	}
	// keep track of the interval of activity too
	Participation activeTime;
	activeTime.setStartTime(participation.getStartTime());
	activeTime.setEndTime(participation.getEndTime());
	activeTime.setIntensity(1);*/
	for (i = 0; i < this->frequencyEstimators.size(); i++)
	{
		frequencyEstimators[i].addParticipationInterval(participation);
	}
	numRatings++;
	latestRatingTime = participation.getEndTime();
}
// returns the number of MovingAverages that try to estimate the current rating
int Candidate::getNumRatingEstimators(void)
{
	return this->ratingEstimators.size();
}
// returns a particular rating estimator
MovingAverage* Candidate::getRatingEstimatorAtIndex(int index)
{
	return &(this->ratingEstimators[index]);
}
// returns the number of MovingAverages that try to estimate how often this song has been listened to recently
int Candidate::getNumFrequencyEstimators(void)
{
	return this->frequencyEstimators.size();
}
// returns a particular frequency estimator
ParticipationMovingAverage* Candidate::getFrequencyEstimatorAtIndex(int index)
{
	return &(this->frequencyEstimators[index]);
}
// returns the moving average that records the exact ratings
RatingMovingAverage* Candidate::getActualRatingHistory(void)
{
	return &(this->actualRatingHistory);
}
Distribution Candidate::getCurrentRating(void)
{
	return this->currentRating;
}
void Candidate::setCurrentRating(Distribution value)
{
	this->currentRating = value;
}
Distribution Candidate::getCurrentRefinedRating(void)
{
	return this->currentRefinedRating;
}
void Candidate::setCurrentRefinedRating(Distribution value)
{
	this->currentRefinedRating = value;
}
bool Candidate::needToUpdateParentPointers(void)
{
	if (this->parentLinksNeeedUpdating)
	{
		parentLinksNeeedUpdating = false;
		return true;
	}
	return false;
}


void Candidate::setDiscoveryDate(DateTime when)
{
	this->discoveryDate = when;
}
// Tells how long it has been since the song was heard. If it's never been heard then it tells how longs it's been since the song was discovered
double Candidate::getIdleDuration(DateTime when)
{
	DateTime latestDate;
	//latestDate = this->actualRatingHistory.getLatestRatingDate();
	ParticipationMovingAverage* frequencies = this->getFrequencyEstimatorAtIndex(0);
	if (frequencies->getNumParticipations() > 0)
		latestDate = frequencies->getLatestDate();
	else
		latestDate = this->discoveryDate;
	return latestDate.timeUntil(when);
}
double Candidate::getAverageRating(void)
{
	return this->actualRatingHistory.getAverageValue();
}

void Candidate::initialize(void)
{
	numRatings = 0;
	/*
#ifdef LONG_SCALE
	// Setup a bunch of moving averages with durations from 15 min to about 4 years
	double currentHalfLife = 900;
#else
	// setup a bunch of moving averages with durations from 1 sec to about a day
	double currentHalfLife = 1;
#endif
	double weightForOldRatings = 1;
	*/
	//const int numAverages = 9;
	const int numAverages = 1;
	int i;
	this->ratingEstimators.resize(numAverages);
	this->frequencyEstimators.resize(numAverages);
	for (i = 0; i < numAverages; i++)
	{
		//frequencyEstimators[i].setHalfLife(currentHalfLife);
		frequencyEstimators[i].setName(Name(this->name.getName() + " (participations)"));
		frequencyEstimators[i].setOwnerName(this->name);
		//frequencyEstimators[i].setOwner(this);
		//currentHalfLife *= 4;
	}
	for (i = 0; i < numAverages; i++)
	{
		//ratingEstimators[i].setWeightForOldRatings(weightForOldRatings);
		ratingEstimators[i].setName(Name(this->name.getName() + " (ratings)"));
		ratingEstimators[i].setOwnerName(this->name);
		//ratingEstimators[i].setOwner(this);
		//weightForOldRatings *= 2;
		//weightForOldRatings *= 4;
	}
	this->actualRatingHistory.setName(Name(this->name.getName() + " actual "));
	this->actualRatingHistory.setOwnerName(this->name);
	//this->actualRatingHistory.setOwner(this);
	//this->actualRatingHistory.setWeightForOldRatings(0);
}

File: chrome/content/algorithm/Candidate.js
/*
    Copyright (C) 2011 Bluejay

    This file is part of Bluejay.

    Bluejay is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Bluejay is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Bluejay.  If not, see <http://www.gnu.org/licenses/>.
*/

/**
 * @class The Candidate class represent item such as a song, artist,
 * genre or other categories. All these items can have ratings.
 */
function Candidate(passedVal) {

	/** public function */
	/** Name */
	this.setName = setName;
	this.getName = getName;
	
	/** unique identifier */
	this.setID = setID;
	
	/**
     * whether this represents an actual music track that can
	 * played, or a genre that can't be played directly
	 */
    this.isPlayable = isPlayable;
    this.setPlayable = setPlayable;
    	
	/** adding parent links */
	this.addParentName = addParentName;
	this.addParent = addParent;
	
	/** adding child links */
	this.addChild = addChild;
	this.getParentNames = getParentNames;
	this.getParents = getParents;
	this.getChildren = getChildren;
	this.getNumChildren = getNumChildren;
	
	/** informing the Candidate whenever the user rates it */
	this.giveRating = giveRating;
	
	/** informing the Candidate when it is listened to */
	this.giveParticipation = giveParticipation;
	
	/**
     * how many MovingAverages there are to estimate its rating.
	 * Currently it's always 1
     */
	this.getNumRatingEstimators = getNumRatingEstimators;
	
	/** get the appropriate MovingAverage */
	this.getRatingEstimatorAtIndex = getRatingEstimatorAtIndex;
	
	/** get the appropriate MovingAverage for different index */
	this.getFrequencyHistory = getFrequencyHistory;
	
    /** get the MovingAverage that stores the ratings */
	this.getRatingHistory = getRatingHistory;
	
	/**
        * the current expected rating, based
        * on data from other Candidates
        */
	this.getCurrentRating = getCurrentRating;
	this.setCurrentRating = setCurrentRating;

       /** the latest date for which a rating was assigned	*/
	this.getLatestRatingDate = getLatestRatingDate;
		
	/** whether the parent pointers are up to date */
	this.needToUpdateParentPointers = needToUpdateParentPointers;
	
	/** the date it was added to your library */
	this.setDiscoveryDate = setDiscoveryDate;
	this.getDiscoveryDate = getDiscoveryDate;
	this.suspectDiscoveryDate = suspectDiscoveryDate;

	/**
        * the duration (in seconds) between the 
        * latest listening and the time at 'when'
        */
	this.getIdleDuration = getIdleDuration;
	this.getDurationSinceLastPlayed = getDurationSinceLastPlayed;
	// the average of all given ratings 
	this.getAverageRating = getAverageRating;
	
	
	/* private functions */
	var initialize;
	
	/** private variables */
	var name = passedVal;
	var ID;
	var playable = true;
	
	/**
     * parentNames is only used at the beginning for storing
	 * names before the parents exist
	 */
	var parentNames = [];
	
	/**
        * 'parents' holds pointers to the parents themselves.
        * It is faster to use than parentNames because removes
        * the need for another dictionary lookup
       */
	var parents = [];
	var children = [];
	var frequencyHistory = new ParticipationMovingAverage();
	var ratingHistory = new RatingMovingAverage();
	/** the latest date at which there was an interaction */
	var latestInteractionDate = new DateTime();
	var lastPlayDate = new DateTime();
	
	/**
        * the rating based on PredictionLinks before incorporating
        * information about the parents
        */
	var currentRating = new Distribution();	
	var latestRatingDate = new DateTime();	
	var parentLinksNeedUpdating = false;
	var discoveryDate;
    
	/** call the constructor */
	if (name) {
    	initialize();
    }
	
	/** public functions */
	
	/**
        * Setting the name of the Candidate must be done before
        * assigning any ratings or participations
        */
	function setName(newName) {
		name = newName.makeCopy();
		/**
        * by default, the ID is the same as the name but
        * may be given a different value
       */
		if (!ID) {
		    ID = name;
		}
		initialize();
	}
	
	/** unique identifier */
	function setID(newID) {
	    ID = newID;
	}
	function getID() {
	    return ID;
    }

    function isPlayable() {
        return playable;
    }
	/** parent/child connections */
	function getName() {
		return name;
	}
	
	/** tell the candidate that it has a parent with this name */
	function addParentName(newName) {
		parentNames.push(newName);
		parentLinksNeedUpdating = true;
	}
	
	/** adding a parent */
	function addParent(newCandidate) {
		parents.push(newCandidate);
	}
	
	/** adding child */
	function addChild(newChild) {
	    playable = false;
		children.push(newChild);
	}
	function setPlayable(canBePlayed) {
	    playable = canBePlayed;
	}
	
	/** get parent names */
	function getParentNames() {
		return parentNames;
	}
	
	/** get parents */
	function getParents() {
		return parents;
	}
	
	/** get children */
	function getChildren() {
		return children;
	}
	
	/** get number of children */
	function getNumChildren() {
	    return children.length;
	}
	
	/** inform the candidate that the user gave it this rating */
	function giveRating(rating)	{
	    // record the rating
		var i = 0;
		/*for (i = 0; i < ratingEstimators.length; i++) {
			ratingEstimators[i].addRating(rating);
		}*/
		ratingHistory.addRating(rating);

		/** update the latest interaction date */
		latestInteractionDate = rating.getDate();
		
		/**
 * If we don't know when we were discovered, then
 * assume we were discovered along with the first
 * rating.  If we think we were discovered after
 * this rating, then it is actually more reasonable
 * to say we were discovered when this rating was created
 */
		if ((!discoveryDate) || (strictlyChronologicallyOrdered(rating.getDate(), discoveryDate))) {
		    discoveryDate = rating.getDate();
		}
	}
	
	/** inform the Candidate that it was listened
        * to during a certain interval
        */
	function giveParticipation(participation) {
		var i = 0;
		/** update the moving averages of the ratings */
		/*for (i = 0; i < frequencyEstimators.length; i++) {
			frequencyEstimators[i].addParticipationInterval(participation);
		}*/
		frequencyHistory.addParticipationInterval(participation);
		lastPlayDate = latestInteractionDate = participation.getEndTime();

		/** 
        * If we don't know when we were discovered, then assume
        * we were discovered along with the first participation.
        * If we think we were discovered after this rating, then
        * it is actually more reasonable to say we were discovered
        * when this rating was created
        */
		if ((!discoveryDate) || (strictlyChronologicallyOrdered(participation.getStartTime(), discoveryDate))) {
		    discoveryDate = participation.getStartTime();
		}
	}
	
	/**
        * returns the number of MovingAverages that try
        * to estimate the current rating
        */
	function getNumRatingEstimators() {
		return ratingEstimators.length;
	}
	
	/** returns a particular rating estimator */
	function getRatingEstimatorAtIndex(index) {
		return ratingEstimators[index];
	}
		
	/**
        * returns the ParticipationMovingAverage
        * that keeps track of listening
        */
	function getFrequencyHistory() {
		return frequencyHistory;
	}
	
	/** returns the moving average that records the ratings */
	function getRatingHistory() {
		return ratingHistory;
	}
	
	/** returns current rating */
	function getCurrentRating() {
		return currentRating;
	}
	
	/** set current rating */
	function setCurrentRating(value, when) {
		currentRating = value;
		latestRatingDate = when;
	}

       /** get the latest date for which a rating was assigned */
	function getLatestRatingDate() {
	    return latestRatingDate;
	}
			
	function needToUpdateParentPointers() {
		if(parentLinksNeedUpdating == true) {
			parentLinksNeedUpdating = false;
			return true;
		}
		return false;
	}
	
	/**
        * set the estimate for the date at which this Candidate was
        * discovered.  Returns true if the estimate is believable,
        * false if the estimate is not believable
        */
	function suspectDiscoveryDate(when) {
		/**
        * Internally we estimate the discovery date based on the
 * first time this song was played or rated.  So, we only
 * use outside suspicion of the discovery date if it does
 * not move the discovery date later
 */
	    if ((!discoveryDate) || (strictlyChronologicallyOrdered(when, discoveryDate))) {
	        this.setDiscoveryDate(when);
	        return true;
        }
        return false;
	}
	/**
 * declares that this Candidate was discovered that the given date
 */
	function setDiscoveryDate(when) {
       /**
        * update the latest date at which there was an interaction
        */
       if (strictlyChronologicallyOrdered(latestInteractionDate, when)) {
       latestInteractionDate = when;
       }
       discoveryDate = when;
	}
	/**
        * get the estimate for the date at which
        * this Candidate was discovered
       */
       function getDiscoveryDate(when) {       /** if discoveryDate is valid... */
       if (discoveryDate) {
           /** ...then return it */
           return discoveryDate;
       }
       else {
           /**
            * If we get here, then we don't have any data
            * suggesting a discovery date.  So, assume it
            * was discovered an extremely long time agon
            */
           return new DateTime();
		}
	}
	
	/** Tells how long it has been since the song was heard or rated */
	function getIdleDuration(when) {
		return latestInteractionDate.timeUntil(when);
	}
	function getDurationSinceLastPlayed(when) {
		return lastPlayDate.timeUntil(when);
	}
	
	function getAverageRating() {
		return ratingHistory.getAverageValue();
	}
	
	function initialize() {
		frequencyHistory.setName(new Name(name.getName() + " (participations)"));
		frequencyHistory.setOwnerName(name);

		ratingHistory.setName(new Name(name.getName() + " (ratings)"));
		ratingHistory.setOwnerName(name);
	}
	
};

